lib/charms/mongodb/v0/upgrade.py

```diff
@@ -0,0 +1,188 @@
+
+from charms.data_platform_libs.v0.upgrade import (
+    ClusterNotReadyError,
+    DataUpgrade,
+    DependencyModel,
+    UpgradeGrantedEvent,
+)
+from charms.mongodb.v0.mongodb import MongoDBConfiguration, MongoDBConnection
+from ops.charm import CharmBase
+from ops.model import ActiveStatus
+from pydantic import BaseModel
+from typing_extensions import override
+
+from config import Config
+
+logger = logging.getLogger(__name__)
+
+WRITE_KEY = "write_value"
+
+# The unique Charmhub library identifier, never change it
+LIBID = "aad46b9f0ddb4cb392982a52a596ec9b"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+
+class MongoDBDependencyModel(BaseModel):
+    """Model for Kafka Operator dependencies."""
+
+    mongod_service: DependencyModel
+    # in future have a mongos service here too
+
+
+class MongoDBUpgrade(DataUpgrade):
+    """Implementation of :class:`DataUpgrade` overrides for in-place upgrades."""
+
+    def __init__(self, charm: CharmBase, **kwargs):
+        super().__init__(charm, **kwargs)
+        self.charm = charm
+
+    @override
+    def pre_upgrade_check(self) -> None:
+        """Verifies that an upgrade can be done on the MongoDB deployment."""
+        default_message = "Pre-upgrade check failed and cannot safely upgrade"
+
+        if self.charm.is_role(Config.Role.SHARD):
+            raise ClusterNotReadyError(
+                message=default_message, cause="Cannot run pre-upgrade check on shards"
+            )
+
+        if not self.is_cluster_healthy():
+            raise ClusterNotReadyError(message=default_message, cause="Cluster is not healthy")
+
+        if not self.is_cluster_able_to_read_write():
+            raise ClusterNotReadyError(message=default_message, cause="Cluster cannot read/write")
+
+        # Future PR - sharding based checks
+
+    @override
+    def build_upgrade_stack(self) -> list[int]:
+        """Builds an upgrade stack, specifying the order of nodes to upgrade.
+
+        TODO Implement in DPE-3940
+        """
+
+    @override
+    def log_rollback_instructions(self) -> None:
+        """Logs the rollback instructions in case of failure to upgrade.
+
+        TODO Implement in DPE-3940
+        """
+
+    @override
+    def _on_upgrade_granted(self, event: UpgradeGrantedEvent) -> None:
+        """Execute a series of upgrade steps.
+
+        TODO Implement in DPE-3940
+        """
+
+    def is_cluster_healthy(self) -> bool:
+        """Returns True if all nodes in the cluster/replcia set are healthy."""
+        if self.charm.is_role(Config.Role.SHARD):
+            logger.debug("Cannot run full cluster health check on shards")
+            return False
+
+        charm_status = self.charm.process_statuses()
+        return self.are_nodes_healthy() and isinstance(charm_status, ActiveStatus)
```

Comment:
The status is more of a UI concept, in a future PR it may be worth having a `process_states` returning a state (enum for ex) which eventually gets mapped to a status in `process_statuses`, so you could rely on it in your conditional branching logic